### PR TITLE
Form focus

### PR DESCRIPTION
**Pull request by mamalos**
An attempt to revive and merge the abandoned PR #268.

### CODE REVIEW

File: CHANGES.txt [hunk 1]
```diff
@@ -86,6 +86,15 @@
...
 
+- Removed deform.js:focusFirstInput()
+  Forms have an optional focus parameter ('on'|'off')
+  'on': if the form contains an input with autofocus='on', that input field 
+        will receive focus.
+  'on': if the form does not contain an input with autofocus='on', the first 
+        input field of the form will receive focus.
+  'off' disables focusing for that form.
+  Default behaviour is 'on' (unchanged).
+
```
Comment (stevepiercy) [inline CHANGES.txt]: 1. This should be moved under the section `2.0.6 (unreleased)`.
2. [Unexpected indentation](https://travis-ci.org/Pylons/deform/jobs/356946946#L595) is causing the docs build to fail. This should be:

```rst
- Removed deform.js:focusFirstInput(). Forms have an optional focus parameter
  ('on'|'off'):

    - 'on': if the form contains an input with autofocus='on', that input
      field will receive focus.
    - 'on': if the form does not contain an input with autofocus='on', the
      first input field of the form will receive focus.
    - 'off' disables focusing for that form.

  Default behaviour is 'on' (unchanged).
```

Test with https://pandoc.org/try/?text=-+Removed+deform.js%3AfocusFirstInput().+Forms+have+an+optional+focus+parameter%0A++(%27on%27%7C%27off%27)%3A%0A%0A++++-+%27on%27%3A+if+the+form+contains+an+input+with+autofocus%3D%27on%27%2C+that+input%0A++++++field+will+receive+focus.%0A++++-+%27on%27%3A+if+the+form+does+not+contain+an+input+with+autofocus%3D%27on%27%2C+the%0A++++++first+input+field+of+the+form+will+receive+focus.%0A++++-+%27off%27+disables+focusing+for+that+form.%0A%0A++Default+behaviour+is+%27on%27+(unchanged).%0A&from=rst&to=html5

File: CHANGES.txt [hunk 2]
```diff
@@ -1,7 +1,7 @@
 2.0.6 (unreleased)
 ------------------
 
-- Nothing changed yet.
+- Form autofocus on first field is now configurable.
```
Comment (stevepiercy) [inline CHANGES.txt]: Suggest adding a reference to this PR, e.g., "See https://github.com/Pylons/deform/pull/365".

**Comment by stevepiercy** [comment]
I think this PR will close #337 and #249 too.

**Comment by stevepiercy** [comment]
I'm working with @mamalos to try to get the functional tests to pass. I updated the instructions in #366 and submitted a couple of other PRs to fix the `.po` files that prevented tox from running. Then I was able to run `tox -e functional3` but not `tox -e functional`. See error below.

I use pyenv for managing my Pythons, and I finally got the tests to run and pass successfully by specifying `pyenv local 3.6.1 2.7.12`. 🎉 

```bash
$ tox -e functional
functional create: /Users/stevepiercy/projects/deform/.tox/functional
ERROR: Error creating virtualenv. Note that some special characters (e.g. ':' and unicode symbols) in paths are not supported by virtualenv. Error details: InvocationError('Failed to get version_info for python2.7: b"pyenv: python2.7: command not found\\n\\nThe `python2.7\' command exists in these Python versions:\\n  2.7.12\\n\\n"',)
________________________________________________________________________________ summary ________________________________________________________________________________
ERROR:   functional: Error creating virtualenv. Note that some special characters (e.g. ':' and unicode symbols) in paths are not supported by virtualenv. Error details: InvocationError('Failed to get version_info for python2.7: b"pyenv: python2.7: command not found\\n\\nThe `python2.7\' command exists in these Python versions:\\n  2.7.12\\n\\n"',)
```

**Comment by stevepiercy** [comment]
w00t! Docs build now passes! https://travis-ci.org/Pylons/deform/jobs/357265162

**Comment by dwt** [comment]
Anything missing to get this merged?

**Comment by mamalos** [comment]
Yep, it's missing me (or maybe you :smiley: ) having some spare time to spend it in understanding and correcting the errors :smiley: . Unfortunately I have been very busy for the last weeks, but I'll try to catch-it-up again this week.

**Comment by mamalos** [comment]
OOoo you killed me...now I have to find time for this! :smiley: 

Unfortunately a veeery busy week is coming, but I'll try catching it in 10 days from now (unless I find some time this weekend).

**Comment by ericof** [comment]
@mamalos As soon as you have it, please ping me and I can take a look

**Comment by stevepiercy** [comment]
@mamalos I am preparing a release of Deform 3.0, now that I have maintainer rights on PyPI, and I think this would be a good feature to include. Would you have some time to rebase your work on master and push to this PR? If not, we can defer to v.3.1. Please let me know. Thank you!

**Comment by mamalos** [comment]
Hi @stevepiercy! Glad you finally have the appropriate rights on PyPI, you're an active member of pyramid, so you more than deserve it (to be honest, I thought you already had such access :-P)!

So, in terms of this infamous PR, I'd be glad to help, but I have the following two problems/constraints:

1) I almost don't remember anything regarding this PR, only faint memories that I've tried to forget (without success).
2) My time is very limited, at least for the next two weeks (yeah, I know how it sounds, but I really am veeeeery pressed at this time).

In terms of (2), I really want to find time contributing to pyramid, because I love it and it has helped me in a few projects so far. I have another [pending PR in cornice](https://github.com/Cornices/cornice/pull/515), for which I'm almost equally ashamed of as well, so I'll try doing both of them in order to "fix" my karma...

In terms of (1), what I need from you is to give me some insight regarding what I need to do. Should I just try to rebase my work on current master and try to see if CI/CD will pass the tests, or do I need to add any fixes to make things work? Because the latter would be very painfully slow from my side **if** it needs javascript work. If so, I'll need more help. If not, I'll try to do my best without bugging you.

**Comment by stevepiercy** [comment]
A rebase of your branch on master, and push to update this PR would be appreciated. Beyond that, I'll take a look, and see what, if anything further, needs to be done.

Because I have write access to this repo and you have opened a PR from your branch against this repo, I can push commits to your branch and update this PR. However I prefer to ask you, the original contributor, to do so, both as a courtesy and to offer you the satisfaction of making a contribution to the FOSS project.

I understand that donating time is not always possible, and I don't expect anyone to do so. I'm grateful when people give their time to make things a little better. Thank you!

**Comment by mamalos** [comment]
Cool! If it's just a rebase to the new master, I'll try to do it tomorrow morning (with a "fresher" mind, that is) and see how it goes, after I'll ve "sharpened" my git/github rebasing skills :).

**Comment by mamalos** [comment]
Hmmm...I tried tiding up a bit and then rebase, but had a few conflicts. Will try again tomorrow

**Comment by mamalos** [comment]
Nice work @stevepiercy! Well done!! You've covered all the steps that I couldn't cover! :+1: 

From my side, I'm perfectly fine with the code (not that I understood everything, since I sort-of-suck at HTML/JS), so I think you can happily integrate it with master.

Thanks for all the help!

**Comment by stevepiercy** [comment]
Thank you @mamalos for jump starting this issue!

**Pull request merged**